### ISSUE TITLE
index: 2 Got: 384 Expected: 640 (again)

### ISSUE THREAD

**Issue by DylanBPY** [ultralytics/yolov5#12722]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

This is the 2nd issue Ive made about this issue. The first time it wasn't so much of an issue because i could just use PyTorch.
Now I am try to use SparseML to make my model a little faster. ONNX is required for SparseML models, so I need to get this working. 

So this is the error that I get:
```
onnxruntime.capi.onnxruntime_pybind11_state.InvalidArgument: [ONNXRuntimeError] : 2 : INVALID_ARGUMENT : Got invalid dimensions for input: images for the following indices
 index: 2 Got: 384 Expected: 640
 Please fix either the inputs or the model.
```

I've tried specifying the size parameter, which does nothing.
I'm, not sure where to go with this, I would like to get this working in the next couple days so that I can release my project as soon as possible.

Here is my the part of my code that matters:
```python
import torch
import cv2

# Load YOLO model
model = torch.hub.load('ultralytics/yolov5', 'custom', path="model.onnx", _verbose=False)
model.conf = 0.50

while True:
    frame = cv2.imread(fr"{path}\Unchecked\{file}") # Image from file
    results = model(frame)
    boxes = results.pandas().xyxy[0]
```

**Comment by glenn-jocher** [comment]
@DylDevs hello! It seems like there might be a mismatch between the expected input size of your ONNX model and the size of the images you are providing. The error indicates that the model is expecting an input size of 640 on a certain dimension, but it's receiving 384 instead.

Here are a few steps you can take to resolve this issue:

1. Ensure that your ONNX model was exported with the correct input size. If you trained your model with a specific input size, you should export it with the same dimensions.

2. Preprocess your input images to match the expected input size of the model. You can resize your images to the required dimensions before passing them to the model.

3. Double-check the ONNX export settings and the documentation to ensure that the model is being used correctly with SparseML.

If you continue to experience issues, please provide more details about how you exported your model to ONNX and any additional code related to the preprocessing of your images. This will help in diagnosing the problem more effectively.

Remember to refer to our documentation for more detailed guidance on model export and usage: https://docs.ultralytics.com/yolov5/

**Comment by DylanBPY** [comment]
I got it working be resizing my inputs to 640x640
However this is not ideal. I would like my images to remain the same size and not need to be messed with

**Comment by glenn-jocher** [comment]
Great to hear you got it working, @DylDevs! If you want to maintain the original image size without resizing, you can consider the following:

1. **Padding**: Instead of resizing, you can pad your images to the required input size with a constant value (like zero-padding). This way, the aspect ratio remains unchanged.

2. **Dynamic Shapes**: Some ONNX runtimes support dynamic input shapes, which allow the model to process varying image sizes. However, this might require re-exporting your model with dynamic axes specified.

3. **Model Modification**: If maintaining the original image size is crucial, you might need to modify the model architecture to accept variable input sizes. This could involve significant changes to the model and is not generally recommended unless necessary.

For now, padding is likely the easiest solution that preserves the original dimensions of your images. Keep in mind that changing the input size can affect the model's performance since it was trained with a specific input resolution in mind.

If you need further assistance, don't hesitate to reach out. Good luck with your project! 🚀

**Comment by DylanBPY** [comment]
1. Padding will not work, the image dimensions must be 640x640, it cant be anything different.

I am not sure how 2 or 3 would work, could you provide some insight on how I do these?'
Also, I would rather not do any sort of image processing, I would like the model to be able work with any size image, like pytorch models.

**Comment by glenn-jocher** [comment]
Understood, @DylDevs. If you want your ONNX model to handle any image size like PyTorch models do, you'll need to look into option 2, which involves dynamic shapes.

Here's a brief overview of how to work with dynamic input shapes in ONNX:

1. **Export with Dynamic Axes**: When exporting your PyTorch model to ONNX, you can specify dynamic axes for your input tensor. This tells the ONNX runtime that the input size can vary. However, not all ONNX runtimes may support dynamic shapes, so you'll need to check compatibility with SparseML.

2. **Runtime Support**: Ensure that the ONNX runtime you are using supports dynamic shapes. You might need to look into the documentation of the ONNX runtime to understand how to work with dynamic shapes.

Unfortunately, without image preprocessing (like padding or resizing), you are limited by the capabilities of the ONNX runtime and the way the model was trained and exported. If the runtime or the model does not support dynamic input shapes, you may not be able to avoid preprocessing.

Since ONNX models are generally optimized for fixed input sizes for performance reasons, handling arbitrary input sizes might require additional steps or compromises. If SparseML and your chosen ONNX runtime do not support dynamic shapes, you may need to stick with the PyTorch model for this flexibility.

For more detailed guidance on exporting models with dynamic axes and using them with ONNX runtimes, please refer to the official ONNX documentation and SparseML's documentation.

**Comment by DylanBPY** [comment]
I would like to just use PyTorch, but SparseML appears to only support ONNX. I also can't find anything related to my issue in the ONNX docs.

**Comment by glenn-jocher** [comment]
I understand the challenge, @DylDevs. SparseML's current support for ONNX models does indeed necessitate the use of ONNX format, which typically requires fixed input dimensions. If the ONNX and SparseML documentation don't provide a solution for dynamic input shapes, you might be facing a limitation of the current tooling.

Here are a couple of suggestions:

1. **Reach out to SparseML**: Consider opening an issue or discussion on the SparseML repository or forums. The maintainers or community may have insights or workarounds for using models with dynamic input sizes.

2. **Custom Implementation**: As a more advanced solution, you could potentially implement custom layers or a wrapper in the ONNX model that handles dynamic resizing within the model itself. This would be quite complex and may require in-depth knowledge of ONNX internals.

3. **Model Architecture**: Some models are inherently more flexible with input sizes. If your current architecture is too rigid, you might explore architectures known for handling variable input dimensions better.

4. **Feedback Loop**: Share your use case and the need for dynamic input support with the SparseML and ONNX communities. If there's enough demand, it could lead to feature development in future releases.

For now, if none of these options are viable and you must use SparseML, you may need to continue with the preprocessing step of resizing your images to the required input size of 640x640. It's not ideal given your requirements, but it might be the only solution given the current tooling constraints.

Keep an eye on updates from SparseML and ONNX, as the landscape for ML tools is constantly evolving, and support for dynamic input sizes may improve in the future.
